File: scripts/generate-spiritual-audio.mjs
#!/usr/bin/env node
/**
 * Spiritual Audio Generator – Gemini 2.5 Flash TTS
 * =================================================
 * Generates calm, meditative chanting audio for Bhagavad Gita slokas.
 *
 * Voice: "Aoede" (UI name: "Devi वाणी")
 * Languages: Sanskrit (chanting), English (meaning), Hindi (अर्थ)
 *
 * Usage:
 *   node scripts/generate-spiritual-audio.mjs --chapter 1 [--verses 1-5] [--lang sanskrit,english,hindi]
 */

import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);
const ROOT = path.resolve(__dirname, '..');

// ─── Config ────────────────────────────────────────────────────────
const API_KEY = process.env.EXPO_PUBLIC_GEMINI_API_KEY
  || process.env.GEMINI_API_KEY
  || (() => {
    try { return fs.readFileSync(path.join(ROOT, '.env'), 'utf8').match(/EXPO_PUBLIC_GEMINI_API_KEY=(.+)/)?.[1]?.trim(); } catch { return null; }
  })();

if (!API_KEY) {
  console.error('❌ No Gemini API key found. Set GEMINI_API_KEY or EXPO_PUBLIC_GEMINI_API_KEY.');
  process.exit(1);
}

const MODEL = 'gemini-3.1-flash-tts-preview';
const VOICE_NAME = 'Aoede'; // Calm, breezy female voice — UI name: "Devi वाणी"
const API_URL = `https://generativelanguage.googleapis.com/v1beta/models/${MODEL}:generateContent?key=${API_KEY}`;
const OUTPUT_DIR = path.join(ROOT, 'public', 'audio');

// ─── Spiritual Prompts ─────────────────────────────────────────────

const SYSTEM_PROMPTS = {
  sanskrit: `You are a sacred Sanskrit scholar and spiritual chanter (Devi वाणी). 
Chant the following Bhagavad Gita shloka in a slow, meditative, deeply devotional manner.
Your voice should evoke the feeling of being in a peaceful temple at dawn.
Pronunciation must be flawless Devanagari Sanskrit with proper chandas (meter).
Pace: Very slow and reverent. Each word should breathe and resonate.
Add natural pauses between half-verses (marked by |).
Do not add any commentary or introduction — chant only the shloka text.`,

  english: `You are a gentle spiritual guide (Devi वाणी) explaining the Bhagavad Gita.
Read the following English translation in a calm, peaceful, meditative voice.
Speak slowly and clearly, as if guiding someone through a quiet meditation.
Your tone should be warm, compassionate, and deeply peaceful.
Do not add any commentary beyond the provided text.
Pace: Slow and contemplative, with natural breathing pauses.`,

  hindi: `आप एक शांत आध्यात्मिक मार्गदर्शक (देवी वाणी) हैं जो भगवद्गीता का हिंदी अनुवाद पढ़ रही हैं।
निम्नलिखित हिंदी अनुवाद को धीमी, शांत, ध्यानमग्न आवाज़ में पढ़ें।
आपकी आवाज़ में करुणा, शांति और आध्यात्मिकता होनी चाहिए।
गति: धीमी और चिंतनशील। प्राकृतिक विराम लें।
केवल दिया गया पाठ पढ़ें, कोई टिप्पणी न जोड़ें।`,
};

// ─── Data Loading ──────────────────────────────────────────────────

function loadVerses(chapter) {
  const gitaPath = path.join(ROOT, 'src', 'data', 'bhagavad-gita.json');
  const data = JSON.parse(fs.readFileSync(gitaPath, 'utf8'));
  const ch = data.chapters.find(c => c.chapter === chapter);
  if (!ch) throw new Error(`Chapter ${chapter} not found`);
  return ch.verses;
}

function loadHindiTranslations() {
  try {
    const hiPath = path.join(ROOT, 'src', 'data', 'languages', 'hi.json');
    return JSON.parse(fs.readFileSync(hiPath, 'utf8'));
  } catch {
    console.warn('⚠️ Hindi translations not found at src/data/languages/hi.json');
    return {};
  }
}

// ─── WAV Helper ────────────────────────────────────────────────────

function createWavHeader(dataLength, sampleRate = 24000, bitsPerSample = 16, channels = 1) {
  const byteRate = sampleRate * channels * (bitsPerSample / 8);
  const blockAlign = channels * (bitsPerSample / 8);
  const header = Buffer.alloc(44);

  header.write('RIFF', 0);
  header.writeUInt32LE(36 + dataLength, 4);
  header.write('WAVE', 8);
  header.write('fmt ', 12);
  header.writeUInt32LE(16, 16);          // PCM
  header.writeUInt16LE(1, 20);           // Audio format (PCM)
  header.writeUInt16LE(channels, 22);
  header.writeUInt32LE(sampleRate, 24);
  header.writeUInt32LE(byteRate, 28);
  header.writeUInt16LE(blockAlign, 32);
  header.writeUInt16LE(bitsPerSample, 34);
  header.write('data', 36);
  header.writeUInt32LE(dataLength, 40);
  return header;
}

// ─── Gemini TTS Generation ─────────────────────────────────────────

async function generateAudio(text, language) {
  const systemPrompt = SYSTEM_PROMPTS[language] || SYSTEM_PROMPTS.english;

  const body = {
    contents: [
      {
        role: 'user',
        parts: [{ text: `${systemPrompt}\n\n---\n\n${text}` }],
      },
    ],
    generationConfig: {
      responseModalities: ['AUDIO'],
      speechConfig: {
        voiceConfig: {
          prebuiltVoiceConfig: {
            voiceName: VOICE_NAME,
          },
        },
      },
    },
  };

  const res = await fetch(API_URL, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body),
  });

  if (!res.ok) {
    const errText = await res.text();
    throw new Error(`Gemini API error ${res.status}: ${errText.substring(0, 300)}`);
  }

  const json = await res.json();
  
  // Extract audio data from response
  const candidate = json.candidates?.[0];
  if (!candidate?.content?.parts) {
    throw new Error('No audio content in response');
  }

  const audioPart = candidate.content.parts.find(p => p.inlineData);
  if (!audioPart) {
    throw new Error('No inline audio data found in response');
  }

  const rawPcm = Buffer.from(audioPart.inlineData.data, 'base64');
  const wavHeader = createWavHeader(rawPcm.length);
  return Buffer.concat([wavHeader, rawPcm]);
}

// ─── WAV → MP3 Conversion (optional via ffmpeg) ──────────────────

async function convertToMp3(wavPath, mp3Path) {
  const { execSync } = await import('child_process');
  try {
    execSync(`ffmpeg -y -i "${wavPath}" -codec:a libmp3lame -qscale:a 2 "${mp3Path}" 2>/dev/null`);
    fs.unlinkSync(wavPath); // clean up wav
    return true;
  } catch {
    console.warn('  ⚠️ ffmpeg not available, keeping WAV format');
    // Rename .wav to .mp3 extension for compatibility (it's still WAV data but works)
    fs.renameSync(wavPath, mp3Path);
    return false;
  }
}

// ─── Main ──────────────────────────────────────────────────────────

async function main() {
  const args = process.argv.slice(2);
  const chapterArg = args.find((a, i) => args[i - 1] === '--chapter') || '1';
  const versesArg = args.find((a, i) => args[i - 1] === '--verses') || null;
  const langArg = args.find((a, i) => args[i - 1] === '--lang') || 'sanskrit,english,hindi';
  const forceRegenerate = args.includes('--force');
  
  const chapter = parseInt(chapterArg);
  const languages = langArg.split(',').map(l => l.trim());
  
  console.log(`\n🙏 Spiritual Audio Generator — Devi वाणी`);
  console.log(`━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
  console.log(`📖 Chapter: ${chapter}`);
  console.log(`🗣️  Voice: Aoede (Devi वाणी)`);
  console.log(`🌐 Languages: ${languages.join(', ')}`);
  console.log(`📂 Output: public/audio/\n`);

  const verses = loadVerses(chapter);
  const hiTranslations = loadHindiTranslations();

  // Parse verse range
  let verseRange = verses;
  if (versesArg) {
    const [start, end] = versesArg.split('-').map(Number);
    verseRange = verses.filter(v => v.verse >= start && v.verse <= (end || start));
  }

  console.log(`📝 Processing ${verseRange.length} verses...\n`);

  // Ensure output directories
  for (const lang of languages) {
    const langDir = path.join(OUTPUT_DIR, lang);
    fs.mkdirSync(langDir, { recursive: true });
  }

  let successCount = 0;
  let failCount = 0;

  for (const verse of verseRange) {
    for (const lang of languages) {
      const verseNum = verse.verse;
      const outFile = `${chapter}_${verseNum}.mp3`;
      const outPath = path.join(OUTPUT_DIR, lang, outFile);

      // Skip if already exists (unless --force)
      if (fs.existsSync(outPath) && !forceRegenerate) {
        console.log(`  ✅ ${lang}/${outFile} — already exists, skipping`);
        continue;
      }

      // Get text for this language
      let text = '';
      if (lang === 'sanskrit') {
        text = verse.sanskrit;
      } else if (lang === 'english') {
        text = verse.translation_english;
      } else if (lang === 'hindi') {
        // Hindi file uses underscore keys: "1_1", "1_2" etc.
        const key = `${chapter}_${verseNum}`;
        text = hiTranslations[key] || '';
        if (!text) {
          console.log(`  ⚠️ ${lang}/${outFile} — no Hindi translation, skipping`);
          continue;
        }
      }

      if (!text) {
        console.log(`  ⚠️ ${lang}/${outFile} — no text found, skipping`);
        continue;
      }

      // Generate audio
      process.stdout.write(`  🎵 Generating ${lang}/${outFile}...`);
      try {
        const wavData = await generateAudio(text, lang);
        const wavPath = outPath.replace('.mp3', '.wav');
        fs.writeFileSync(wavPath, wavData);
        await convertToMp3(wavPath, outPath);
        
        const sizeMB = (fs.statSync(outPath).size / (1024 * 1024)).toFixed(2);
        console.log(` ✅ (${sizeMB} MB)`);
        successCount++;
      } catch (err) {
        console.log(` ❌ ${err.message.substring(0, 200)}`);
        failCount++;
        // If rate limited, wait longer
        if (err.message.includes('429') || err.message.includes('quota')) {
          console.log('  ⏳ Rate limited, waiting 30s...');
          await new Promise(r => setTimeout(r, 30000));
        }
      }

      // Rate limiting: pause 7s between requests (10 req/min quota)
      await new Promise(r => setTimeout(r, 7000));
    }
  }

  console.log(`\n━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━`);
  console.log(`✅ Generated: ${successCount} | ❌ Failed: ${failCount}`);
  console.log(`📂 Files saved to: public/audio/`);
  console.log(`\n🙏 Hari Om Tat Sat\n`);
}

main().catch(err => {
  console.error('Fatal error:', err);
  process.exit(1);
});

File: scripts/generate_spiritual_audio.py
#!/usr/bin/env python3
"""
Generate authentic spiritual chanting audio for Bhagavad Gita verses
using Google Gemini's TTS capabilities.

Usage:
  python3 scripts/generate_spiritual_audio.py --chapter 1

This script generates:
  1. Sanskrit chanting (meditative tone) -> assets/audio/sanskrit/{ch}_{v}.mp3
  2. English meaning narration (spiritual tone) -> assets/audio/english/{ch}_{v}.mp3
"""

import os
import sys
import json
import time
import argparse
import struct
import io

# Try to import google.genai
try:
    from google import genai
    from google.genai import types
except ImportError:
    print("ERROR: google-genai package not installed.")
    print("Run: pip install google-genai")
    sys.exit(1)

# ─── Config ──────────────────────────────────────────────────────────
SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
PROJECT_DIR = os.path.dirname(SCRIPT_DIR)
DATA_FILE = os.path.join(PROJECT_DIR, "src", "data", "bhagavad-gita.json")
SANSKRIT_OUT = os.path.join(PROJECT_DIR, "assets", "audio", "sanskrit")
ENGLISH_OUT = os.path.join(PROJECT_DIR, "assets", "audio", "english")

# Gemini model for TTS
TTS_MODEL = "gemini-2.5-flash-preview-tts"

# Rate limiting
DELAY_BETWEEN_REQUESTS = 2  # seconds


def load_chapter_data(chapter_num: int) -> dict:
    """Load verse data for a specific chapter."""
    with open(DATA_FILE, "r", encoding="utf-8") as f:
        data = json.load(f)

    for ch in data["chapters"]:
        if ch["chapter"] == chapter_num:
            return ch

    raise ValueError(f"Chapter {chapter_num} not found in data file.")


def clean_sanskrit(text: str) -> str:
    """Clean Sanskrit text for chanting - remove verse numbers and pipes."""
    import re
    text = re.sub(r'\|\|[\d\-]+\|\|', '', text)
    text = re.sub(r'\|+', '', text)
    text = text.replace('\\n', '\n')
    return text.strip()


def clean_english(text: str) -> str:
    """Clean English text for narration."""
    import re
    text = re.sub(r'^[\d.]+\s*', '', text)
    text = re.sub(r'Chapter \d+, Verse \d+[.,]?\s*', '', text, flags=re.IGNORECASE)
    text = text.replace('"', '').replace('"', '').replace('"', '')
    return text.strip()


def convert_pcm_to_wav(pcm_data: bytes, sample_rate: int = 24000, channels: int = 1, sample_width: int = 2) -> bytes:
    """Convert raw PCM data to WAV format."""
    wav_buffer = io.BytesIO()
    
    # WAV header
    data_size = len(pcm_data)
    wav_buffer.write(b'RIFF')
    wav_buffer.write(struct.pack('<I', 36 + data_size))
    wav_buffer.write(b'WAVE')
    wav_buffer.write(b'fmt ')
    wav_buffer.write(struct.pack('<I', 16))  # chunk size
    wav_buffer.write(struct.pack('<H', 1))   # PCM format
    wav_buffer.write(struct.pack('<H', channels))
    wav_buffer.write(struct.pack('<I', sample_rate))
    wav_buffer.write(struct.pack('<I', sample_rate * channels * sample_width))  # byte rate
    wav_buffer.write(struct.pack('<H', channels * sample_width))  # block align
    wav_buffer.write(struct.pack('<H', sample_width * 8))  # bits per sample
    wav_buffer.write(b'data')
    wav_buffer.write(struct.pack('<I', data_size))
    wav_buffer.write(pcm_data)
    
    return wav_buffer.getvalue()


def wav_to_mp3(wav_data: bytes, output_path: str):
    """Convert WAV data to MP3 using ffmpeg or just save as WAV if ffmpeg unavailable."""
    import subprocess
    import tempfile
    
    # Write WAV to temp file
    with tempfile.NamedTemporaryFile(suffix='.wav', delete=False) as tmp_wav:
        tmp_wav.write(wav_data)
        tmp_wav_path = tmp_wav.name
    
    try:
        # Try ffmpeg conversion
        result = subprocess.run(
            ['ffmpeg', '-y', '-i', tmp_wav_path, '-codec:a', 'libmp3lame', '-qscale:a', '4', output_path],
            capture_output=True, text=True, timeout=30
        )
        if result.returncode != 0:
            raise RuntimeError(f"ffmpeg failed: {result.stderr}")
    except (FileNotFoundError, RuntimeError):
        # Fallback: save as WAV with .mp3 extension (expo-av handles both)
        print("    ⚠️  ffmpeg not found, saving as WAV (compatible with expo-av)")
        with open(output_path, 'wb') as f:
            f.write(wav_data)
    finally:
        os.unlink(tmp_wav_path)


def generate_audio(client, text: str, voice_name: str, output_path: str, style_prompt: str = "", retries: int = 3):
    """Generate audio using Gemini TTS and save to file with retries."""
    
    full_prompt = f"{style_prompt}\n\n{text}" if style_prompt else text
    
    for attempt in range(retries):
        try:
            response = client.models.generate_content(
                model=TTS_MODEL,
                contents=full_prompt,
                config=types.GenerateContentConfig(
                    response_modalities=["AUDIO"],
                    speech_config=types.SpeechConfig(
                        voice_config=types.VoiceConfig(
                            prebuilt_voice_config=types.PrebuiltVoiceConfig(
                                voice_name=voice_name,
                            )
                        )
                    ),
                ),
            )
            
            # Extract audio data
            if not response.candidates or not response.candidates[0].content.parts:
                print(f"    ⚠️ Warning: Empty response from Gemini (Attempt {attempt+1}/{retries})")
                time.sleep(DELAY_BETWEEN_REQUESTS * 2)
                continue
                
            audio_data = response.candidates[0].content.parts[0].inline_data.data
            mime_type = response.candidates[0].content.parts[0].inline_data.mime_type
            
            # Gemini returns raw PCM audio at 24kHz
            if 'pcm' in mime_type.lower() or 'raw' in mime_type.lower() or 'L16' in mime_type:
                wav_data = convert_pcm_to_wav(audio_data, sample_rate=24000)
            else:
                # If it returns WAV or another format directly
                wav_data = audio_data
            
            # Convert to MP3
            wav_to_mp3(wav_data, output_path)
            
            return True
            
        except Exception as e:
            print(f"    ❌ Error: {e} (Attempt {attempt+1}/{retries})")
            time.sleep(DELAY_BETWEEN_REQUESTS * 2)
            
    return False


def main():
    parser = argparse.ArgumentParser(description="Generate spiritual chanting audio for Gita verses")
    parser.add_argument("--chapter", type=int, required=True, help="Chapter number to generate audio for")
    parser.add_argument("--start-verse", type=int, default=1, help="Starting verse (default: 1)")
    parser.add_argument("--end-verse", type=int, default=0, help="Ending verse (default: all)")
    parser.add_argument("--sanskrit-only", action="store_true", help="Generate only Sanskrit chanting")
    parser.add_argument("--english-only", action="store_true", help="Generate only English narration")
    parser.add_argument("--dry-run", action="store_true", help="Print what would be generated without calling API")
    args = parser.parse_args()

    # Load API key from .env
    env_path = os.path.join(PROJECT_DIR, ".env")
    api_key = None
    if os.path.exists(env_path):
        with open(env_path) as f:
            for line in f:
                if line.startswith("EXPO_PUBLIC_GEMINI_API_KEY="):
                    api_key = line.split("=", 1)[1].strip().strip('"').strip("'")
    
    if not api_key:
        print("ERROR: No Gemini API key found in .env file.")
        print("Set EXPO_PUBLIC_GEMINI_API_KEY in your .env file.")
        sys.exit(1)

    # Initialize client
    client = genai.Client(api_key=api_key)

    # Load data
    chapter_data = load_chapter_data(args.chapter)
    verses = chapter_data["verses"]
    
    end_verse = args.end_verse if args.end_verse > 0 else chapter_data["verses_count"]
    verses = [v for v in verses if args.start_verse <= v["verse"] <= end_verse]

    print(f"\n🕉️  Generating spiritual audio for Chapter {args.chapter}")
    print(f"   Verses: {args.start_verse} to {end_verse} ({len(verses)} verses)")
    print(f"   Sanskrit: {'Yes' if not args.english_only else 'No'}")
    print(f"   English:  {'Yes' if not args.sanskrit_only else 'No'}")
    print()

    # Ensure output directories exist
    os.makedirs(SANSKRIT_OUT, exist_ok=True)
    os.makedirs(ENGLISH_OUT, exist_ok=True)

    # Style prompts
    sanskrit_style = (
        "Chant the following Sanskrit verse from the Bhagavad Gita in a slow, reverent, "
        "meditative tone. Use traditional Vedic chanting rhythm. "
        "Do NOT say any chapter or verse numbers. "
        "Just chant the sacred Sanskrit text with devotion and clarity."
    )

    english_style = (
        "Read the following English translation of a Bhagavad Gita verse in a calm, "
        "spiritual, and contemplative narration style. Speak slowly and clearly, "
        "as if guiding someone in deep meditation. "
        "Do NOT mention any chapter or verse numbers."
    )

    success_count = 0
    fail_count = 0
    
    for verse in verses:
        v_num = verse["verse"]
        sanskrit_text = clean_sanskrit(verse["sanskrit"])
        english_text = clean_english(verse["translation_english"])
        
        print(f"  📿 Chapter {args.chapter}, Verse {v_num}")
        
        if args.dry_run:
            print(f"      Sanskrit: {sanskrit_text[:60]}...")
            print(f"      English:  {english_text[:60]}...")
            continue

        # Generate Sanskrit chanting
        if not args.english_only:
            sanskrit_path = os.path.join(SANSKRIT_OUT, f"{args.chapter}_{v_num}.mp3")
            print(f"      🔊 Sanskrit chanting...", end=" ", flush=True)
            
            if generate_audio(client, sanskrit_text, "Kore", sanskrit_path, sanskrit_style):
                file_size = os.path.getsize(sanskrit_path)
                print(f"✅ ({file_size // 1024} KB)")
                success_count += 1
            else:
                fail_count += 1
            
            time.sleep(DELAY_BETWEEN_REQUESTS)

        # Generate English narration
        if not args.sanskrit_only:
            english_path = os.path.join(ENGLISH_OUT, f"{args.chapter}_{v_num}.mp3")
            print(f"      🔊 English narration...", end=" ", flush=True)
            
            if generate_audio(client, english_text, "Puck", english_path, english_style):
                file_size = os.path.getsize(english_path)
                print(f"✅ ({file_size // 1024} KB)")
                success_count += 1
            else:
                fail_count += 1
            
            time.sleep(DELAY_BETWEEN_REQUESTS)

    print(f"\n{'='*50}")
    print(f"  ✅ Successful: {success_count}")
    print(f"  ❌ Failed: {fail_count}")
    print(f"{'='*50}")
    
    if not args.dry_run:
        # Calculate total size
        total_size = 0
        for d in [SANSKRIT_OUT, ENGLISH_OUT]:
            for f in os.listdir(d):
                if f.startswith(f"{args.chapter}_"):
                    total_size += os.path.getsize(os.path.join(d, f))
        print(f"  💾 Total size for Chapter {args.chapter}: {total_size / (1024*1024):.1f} MB")


if __name__ == "__main__":
    main()
